kvserver: requests that acquire repl locks should use read-write path

arulajmani opened the pull request:
First commit from https://github.com/cockroachdb/cockroach/pull/110201

Previously, {Get,Scan,ReverseScan}Requests all used the read-only
execution path. Things aren't so simple anymore, now that we want these
requests to be able to acquire replicated locks, which means they need
to go through raft (and therefore the read-write execution path). This
patch achieves exactly that.

Informs https://github.com/cockroachdb/cockroach/issues/100193

Release note: None

arulajmani committed:
kvserver: requests that acquire repl locks should use read-write path

Previously, {Get,Scan,ReverseScan}Requests all used the read-only
execution path. Things aren't so simple anymore, now that we want these
requests to be able to acquire replicated locks, which means they need
to go through raft (and therefore the read-write execution path). This
patch achieves exactly that.

Informs #100193

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110279)
<!-- Reviewable:end -->

nvanbenschoten commented on pkg/kv/kvpb/api.go:
```diff
@@ -1354,9 +1370,17 @@ func flagForLockStrength(l lock.Strength) flag {
+	if l == lock.None {
+		return isRead
+	}
+	return isWrite
```
Should this consult the `durabilityType`?

arulajmani replied:
Oops, I was considering a different approach which I decided to backtrack on. Didn't press undo enough times :p

arulajmani committed:
kv: add lock durability options to the KV client API

This patch adds lock durability flags to GetFor{Update,Share},
ScanFor{Update,Share}, and ReverseScanFor{Update,Share}, using which
users of the KV client API can express lock durability goals. These
then map to either replicated or unreplicated locks.

This patch then modifies existing tests to make use of replicated locks
in a few places.

Epic: none

Release note: None

arulajmani committed:
kvserver: add a basic shared locks test

This patch adds a very basic shared locks test using the KV client
API. While here, we also take the opportunity to extend this test
for replicated locks.

Epic: none

Release note: None

arulajmani reviewed (commented):
@nvanbenschoten this should be good for a look now. The client level test for shared locks which prompted the latching changes. I'll spend a bit more time thinking about other tests we may want to add, either in this patch or as followups, but if any come to mind let me know! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-:-NeVIMYtCZT38itX2X9N:bgebeoq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
@nvanbenschoten I've updated this patch with everything we spoke about over call/on slack. Should be good for another look! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-:-Net1360CxeWf3YCiqD0:b-e3t6mo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @miretskiy, and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 8 of 9 files at r3, 6 of 6 files at r4, 29 of 29 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110279)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @cucaroach, and @miretskiy)

___
*[`pkg/kv/db.go` line 521 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net5xfp6-VS-Y1iObVD:-Net5xfp6-VS-Y1iObVE:b-rtacrb) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/db.go#L521)):*
> ```Go
> // key can be either a byte slice or a string.
> func (db *DB) Scan(ctx context.Context, begin, end interface{}, maxRows int64) ([]KeyValue, error) {
> 	return db.scan(ctx, begin, end, maxRows, false /* isReverse */, kvpb.NonLocking, kvpb.CONSISTENT, kvpb.Invalid)
> ```

The direct use of `kvpb.Invalid` seems fine. There's also similar cases where we just pass in `0` in these cases. No preference from me.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go` line 87 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net2r0F1fbSrfjnaDrI:-Net2r0F1fbSrfjnaDrJ:b-bdx35y) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go#L87)):*
> ```Go
> ) (*kvpb.BatchResponse, *kvpb.Error) {
> 	for _, ru := range ba.Requests {
> ```

nit: stray line?
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go` line 101 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net2wfi67oCUWRzq5wL:-Net2wfi67oCUWRzq5wM:bussbix) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go#L101)):*
> ```Go
> 		// Requests that do not perform intent writes use the read sequence number.
> 		// Notably, this includes Get/Scan/ReverseScan requests that acquire
> 		// replicated locks, even though thye go through raft.
> ```

s/thye/they/
___
*[`pkg/kv/kvnemesis/applier.go` line 289 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net6QyK3FvVSW6w-udQ:-Net6QyK3FvVSW6w-udR:b4bndm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvnemesis/applier.go#L289)):*
> ```Go
> 			dur := kvpb.BestEffort
> 			if o.ForUpdate {
> 				(*kv.Batch).GetForUpdate(b, o.Key, dur)
> ```

Here and below, should we just be calling the method on `b` directly, like `b.GetForUpdate(o.Key, dur)`?
___
*[`pkg/kv/kvpb/api.go` line 1386 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net39BO9FvaX0ik-p5x:-Net39BO9FvaX0ik-p5y:bvhnmq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvpb/api.go#L1386)):*
> ```Go
> 
> func flagForIsWrite(l lock.Strength, d lock.Durability) flag {
> 	if l == lock.None || d == lock.Unreplicated {
> ```

Given that the `KeyLockingDurability` can only be set if `KeyLockingStrength` is non-None, it doesn't feel like this function should need to take a strength argument at all. It could be as simple as:
```go
func flagForLockDurability(d lock.Durability) flag {
    if d == lock.Unreplicated {
        return isWrite
    }
    return 0
}
```
___
*[`pkg/kv/kvpb/api.go` line 1208 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetaUBKDWvFCuLa0MxB:-NetaUBKDWvFCuLa0MxC:bfqmzoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvpb/api.go#L1208)):*
> ```Go
> 
> // NewGet returns a Request initialized to get the value at key. No lock is
> // acquired A lock on the key, even if it exists.
> ```

"No lock is acquired A lock on the key".
___
*[`pkg/kv/kvpb/api.go` line 1348 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetabRt0rSMoEBtM8Q6:-NetabRt0rSMoEBtM8Q7:b-a5jhpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvpb/api.go#L1348)):*
> ```Go
> 
> // NewScan returns a Request initialized to scan from start to end keys. No
> // locks will be acquired on each of the resulting keys.
> ```

nit here and below: `s/on each of the/on the/`
___
*[`pkg/kv/kvpb/api.go` line 1410 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetaljM1B8hIR8ADW0D:-NetaljM1B8hIR8ADW0E:b-nxg8mv) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvpb/api.go#L1410)):*
> ```Go
> 
> // KeyLockingDurabilityType is used to describe the durability goals of per-key
> // locks acquired by Get, Scan, and ReverseScan requests.
> ```

"by locking Get, ..."
___
*[`pkg/kv/kvpb/api.go` line 1431 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Netb6hb4Y0Fq7er2TWz:-Netb6hb4Y0Fq7er2TX-:b2wa76w) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvpb/api.go#L1431)):*
> ```Go
> 	// transactions that require locks for correctness (read: read-committed,
> 	// snapshot isolation).
> 	GuaranteedDurability
> ```

Do we want to mention that these locks also prevent writes at or below the commit timestamp of the holder after the lock has been released? So the read cannot skew between the mvcc time that it is originally performed and the mvcc time that the lock holder commits. There's likely a more formal way to express this.
___
*[`pkg/kv/kvpb/api.go` line 1448 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetbiF0CV5wP-n9ojYH:-NetbiF0CV5wP-n9ojYI:b-ppana7) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvpb/api.go#L1448)):*
> ```Go
> 
> func scanLockDurability(dur KeyLockingDurabilityType) lock.Durability {
> 	switch dur {
> ```

Do we want to add a case for `Invalid` with a better panic message?
___
*[`pkg/kv/kvserver/replica_command.go` line 3280 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Netc9IE9FkX6ym99oJc:-Netc9IFEOinX2F8Mn9C:b611ikb) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvserver/replica_command.go#L3280)):*
> ```Go
> // on the locking mode of the read being respected, remove the forUpdate param
> // and perform locking reads for all callers.
> // TODO(XXX): Should we address this TODO? I'm not sure what the transaction
> ```

Addressing this TODO doesn't seem particularly urgent, but if XXX is feeling motivated, XXX should feel free to.
___
*[`pkg/kv/kvserver/batcheval/declare.go` line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NeiHo35-x__rDJlv3Hr:-NeiHo35-x__rDJlv3Hs:bt0yurc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8a752d1891995d5bc67c109e6f1acfee21f6d9d/pkg/kv/kvserver/batcheval/declare.go#L69)):*
> ```Go
> 	// on the request and the evaluation path they take. We handle lock/latch span
> 	// declaration for them along with other read-only requests.
> 	if ok || kvpb.IsReadOnly(req) {
> ```

There are other calls to `IsReadOnly` that also need to change. For example, in `txnSeqNumAllocator`. 
___
*[`pkg/kv/kvserver/batcheval/declare.go` line 90 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net3jIn2uq5r3nUaQY4:-Net3jIn2uq5r3nUaQY5:btutpxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/declare.go#L90)):*
> ```Go
> 		timestamp.Forward(in.GlobalLimit)
> 	} else {
> 		str = lock.Intent
> ```

nit: consider putting this in an `else` branch of the LockingReadRequest check. Up to your coding style preferences.
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 109 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net3xXlCIg28Uh5TfxR:-Net3xXlCIg28Uh5TfxS:b-bhpypz) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L109)):*
> ```Go
> // acquireLocksOnKeys acquires locks on each of the keys in the result of a
> // {,Reverse}ScanRequest. The locks are held by the specified transaction with
> // the supplied locks strength. Best-effort locks are held in unreplicated
> ```

"in an unreplicated"
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 115 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net5-VODS_PUFizSHZL:-Net5-VODS_PUFizSHZM:b-j54dtm) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L115)):*
> ```Go
> // lock on one of the keys. In such cases, a LockConflictError is returned to
> // the caller. However, if locks have been successfully acquired on each of the
> // keys, the provided result.Result is mutated accordingly.
> ```

This is kind of an awkward contract. Should we just return the `[]roachpb.LockAcquisition`? That would parallel the contract of `acquireLockOnKey`. 
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 121 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net4VTC1wkkQrcgqP2V:-Net4VTC1wkkQrcgqP2W:byqh0h1) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L121)):*
> ```Go
> 	res *result.Result,
> 	txn *roachpb.Transaction,
> 	keyLocking lock.Strength,
> ```

Out of curiosity, why the `s/str/keyLocking/`? Should we do the same with `dur`?
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 150 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net4usc5hCjcHBn4HH6:-Net4usc5hCjcHBn4HH7:b7r75bc) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L150)):*
> ```Go
> 			acq, err := acquireLockOnKey(ctx, readWriter, txn, keyLocking, dur, k)
> 			if err != nil {
> 				res.Local.AcquiredLocks = res.Local.AcquiredLocks[:]
> ```

Was this supposed to be `res.Local.AcquiredLocks[:i]`?
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 166 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net4ddq5Kl2nl0S7-ke:-Net4ddq5Kl2nl0S7-kf:b-bhpypz) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L166)):*
> ```Go
> // acquireLockOnKey acquires a lock on the specified key. The lock is acquired
> // with the supplied lock strength and are held by the specified transaction.
> // Best-effort locks are held in unreplicated fashion; locks that need
> ```

"in an unreplicated"
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 168 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net4hU75aNemXZSKuW9:-Net4hU75aNemXZSKuWA:bkinsd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L168)):*
> ```Go
> // Best-effort locks are held in unreplicated fashion; locks that need
> // guaranteed durability are replicated. The resultant lock acquisition struct
> // is returned, which the caller may accumulate in its result set.
> ```

s/may/must/
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 194 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net5NMdANhlZSga4hWZ:-Net5NMdANhlZSga4hW_:b-vybkvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L194)):*
> ```Go
> 		dur = lock.Unreplicated
> 	}
> 	acq := roachpb.MakeLockAcquisition(txn, key, dur, str)
> ```

nit: should this call and the `return acq, nil` be moved after the switch statement?
___
*[`pkg/kv/db_test.go` line 115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetcS9m7EEMMDJ_nF25:-NetcS9m7EEMMDJ_nF26:b4r169i) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/db_test.go#L115)):*
> ```Go
> 	s, db := setup(t)
> 	defer s.Stopper().Stop(context.Background())
> 	testutils.RunTrueAndFalse(t, "durability-guaranteed", func(t *testing.T, durabilityGuaranteed bool) {
> ```

nit: add a new line above this to match `TestDB_GetForShare`.
___
*[`pkg/kv/db_test.go` line 381 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetcXLmB6MMkel3ovAK:-NetcXLmB6MMkel3ovAL:bogwcty) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/db_test.go#L381)):*
> ```Go
> 	defer log.Scope(t).Close(t)
> 	s, db := setup(t)
> 	ctx := context.Background()
> ```

Want to pull out a `ctx` in the tests above as well?
___
*[`pkg/kv/kvserver/client_replica_test.go` line 4992 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetdHqQAwtwfyz6yzsi:-NetdHqQAwtwfyz6yzsj:b-zi8skc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7664df138967f920b4401f64a180ac95ae742a4/pkg/kv/kvserver/client_replica_test.go#L4992)):*
> ```Go
> 		var mu struct {
> 			syncutil.Mutex
> 			numFinalized int
> ```

`atomic.Int64`?
___
*[`pkg/kv/kvserver/client_replica_test.go` line 5007 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Netd8oZ0ch5RfGwAElq:-Netd8oZ0ch5RfGwAElr:b-4xnp3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7664df138967f920b4401f64a180ac95ae742a4/pkg/kv/kvserver/client_replica_test.go#L5007)):*
> ```Go
> 		}()
> 
> 		mu.Lock()
> ```

Should we sleep a bit here to give the GetForUpdate request time to get blocked? Even as small as 10ms? Kind of like in `TestTxnWaitPolicies`.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
TFTR, RFAL. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-:-Ney4T5H4-feuC8cl1eM:bmu8ofu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @miretskiy, and @nvanbenschoten)

___
*[`pkg/kv/db.go` line 521 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net5xfp6-VS-Y1iObVD:-NexnZ9G2L4NR0W3A4v5:b-qrgxhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/db.go#L521)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The direct use of `kvpb.Invalid` seems fine. There's also similar cases where we just pass in `0` in these cases. No preference from me.
</blockquote></details>

Yeah, I'm not entirely sure what I prefer here either. On one hand, `Invalid` makes things explicit at the caller. On the other hand, I've to explain what "Invalid" means at the enum definition. I'll defer to inertia, I guess.
___
*[`pkg/kv/kvnemesis/applier.go` line 289 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net6QyK3FvVSW6w-udQ:-NexoGJkEtPzsGq6yLZ7:b-afnw7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvnemesis/applier.go#L289)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Here and below, should we just be calling the method on `b` directly, like `b.GetForUpdate(o.Key, dur)`?
</blockquote></details>

Done; not entirely sure why it was written like this in the first place.
___
*[`pkg/kv/kvpb/api.go` line 1386 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net39BO9FvaX0ik-p5x:-Nexb7BrAc9deXbgcu2w:b-kyxg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvpb/api.go#L1386)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Given that the `KeyLockingDurability` can only be set if `KeyLockingStrength` is non-None, it doesn't feel like this function should need to take a strength argument at all. It could be as simple as:
```go
func flagForLockDurability(d lock.Durability) flag {
    if d == lock.Unreplicated {
        return isWrite
    }
    return 0
}
```
</blockquote></details>

Good point. Done.  

PS: (I think you inverted the condition above)
___
*[`pkg/kv/kvpb/api.go` line 1431 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Netb6hb4Y0Fq7er2TWz:-NextFI01tY-RTc3eYTF:b88knsb) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvpb/api.go#L1431)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we want to mention that these locks also prevent writes at or below the commit timestamp of the holder after the lock has been released? So the read cannot skew between the mvcc time that it is originally performed and the mvcc time that the lock holder commits. There's likely a more formal way to express this.
</blockquote></details>

Added some words. PTAL.
___
*[`pkg/kv/kvserver/replica_command.go` line 3280 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Netc9IE9FkX6ym99oJc:-NexbjQO1aYbQ5iKB2zk:bngqy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvserver/replica_command.go#L3280)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Addressing this TODO doesn't seem particularly urgent, but if XXX is feeling motivated, XXX should feel free to.
</blockquote></details>

XXX is going to see some natural salt mines tomorrow, so maybe in October :) 
___
*[`pkg/kv/kvserver/batcheval/declare.go` line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NeiHo35-x__rDJlv3Hr:-NexchU6BD8u7SAiqDiX:bey3yob) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8a752d1891995d5bc67c109e6f1acfee21f6d9d/pkg/kv/kvserver/batcheval/declare.go#L69)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

There are other calls to `IsReadOnly` that also need to change. For example, in `txnSeqNumAllocator`.
</blockquote></details>

(from before the recent push -- we ended up discussing this offline). 
___
*[`pkg/kv/kvserver/batcheval/declare.go` line 90 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net3jIn2uq5r3nUaQY4:-NexdLKw1kgLg3RZ8f2D:bmlnq3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/declare.go#L90)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: consider putting this in an `else` branch of the LockingReadRequest check. Up to your coding style preferences.
</blockquote></details>

Minor preference for this, so I'll keep.
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 115 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net5-VODS_PUFizSHZL:-NexgR7Q-rouME9tjz9h:bbh8dcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L115)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is kind of an awkward contract. Should we just return the `[]roachpb.LockAcquisition`? That would parallel the contract of `acquireLockOnKey`.
</blockquote></details>

Yeah, makes sense. I'd rewritten these a few different ways before I finally settled on the contract for `acquireLockOnKey`. Should've unified it with `acquireLocksOnKeys`. Done now.
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 121 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net4VTC1wkkQrcgqP2V:-Nexdh2a6wW7MMEP6w6A:bfcpb4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L121)):*
> Out of curiosity, why the `s/str/keyLocking/`? Should we do the same with `dur`?

I built this on top of my first PR, which hadn't renamed `s/keyLocking/keyLockingStrength` on batch requests. It also didn't prefix durability with `keyLocking`. `s/str/keyLocking/` was prompted by the fields we were passing in at the caller.

I'm actually going to walk back this change. `keyLockingStrength` and `keyLockingDurability` feels needlessly verbose here.
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 150 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net4usc5hCjcHBn4HH6:-NexkTYaE3A24J1aZ-na:ba0jpdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L150)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Was this supposed to be `res.Local.AcquiredLocks[:i]`?
</blockquote></details>

I was just clearing the entire thing because if there's a `LockConflictError` then none of the locks are acquired. Is that not so?
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 166 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net4ddq5Kl2nl0S7-ke:-Nexke-r17hIFaQbzV78:b-cut7fz) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L166)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"in an unreplicated"
</blockquote></details>

Ended up removing this entirely -- this comment was referencing "best effort" and "guaranteed", which we've since walked back on.
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 194 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net5NMdANhlZSga4hWZ:-NexlRu63WMDw43QLiR9:b-uannq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L194)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: should this call and the `return acq, nil` be moved after the switch statement?
</blockquote></details>

Yeah, done.
___
*[`pkg/kv/db_test.go` line 381 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetcXLmB6MMkel3ovAK:-NexrDLC-8vWuREN9e1a:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/db_test.go#L381)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Want to pull out a `ctx` in the tests above as well?
</blockquote></details>

Done. 
___
*[`pkg/kv/kvserver/client_replica_test.go` line 4992 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetdHqQAwtwfyz6yzsi:-Ney1tfF2myelxissnGC:bquu2jw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7664df138967f920b4401f64a180ac95ae742a4/pkg/kv/kvserver/client_replica_test.go#L4992)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`atomic.Int64`?
</blockquote></details>

I thought we'd need to wrap both the increments and `txn1.Commit()` and `txn2.Rollback()` inside of the mutex, because otherwise the `GetForUpdate` might get unblocked and read the wrong value for `numFinalized`. But that also makes this checking `numFinalized` bit kind of stupid. Should we just remove it?
___
*[`pkg/kv/kvserver/client_replica_test.go` line 5007 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Netd8oZ0ch5RfGwAElq:-Nexy2pO46Vk-DYG40k5:b-i47wym) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7664df138967f920b4401f64a180ac95ae742a4/pkg/kv/kvserver/client_replica_test.go#L5007)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we sleep a bit here to give the GetForUpdate request time to get blocked? Even as small as 10ms? Kind of like in `TestTxnWaitPolicies`.
</blockquote></details>

Good call. Done. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 33 of 33 files at r7, 30 of 30 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110279)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @cucaroach, and @miretskiy)

___
*[`pkg/kv/kvnemesis/applier.go` line 289 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net6QyK3FvVSW6w-udQ:-NeycSBJAlHMGf1hGrzf:bm3v743) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvnemesis/applier.go#L289)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Done; not entirely sure why it was written like this in the first place.
</blockquote></details>

If we're assigning the method reference to a variable, then this avoids an allocation to close over `b`. But since we're now just calling the method directly, this is unnecessary.

We'll want to make this change down below as well.
___
*[`pkg/kv/kvnemesis/applier.go` line 445 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Neycme66bbT4llmCaqK:-Neycme66bbT4llmCaqL:b79tnwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cf34d617c8b3e63803bca897cb6da412ac613c0/pkg/kv/kvnemesis/applier.go#L445)):*
> ```Go
> 		switch subO := o.Ops[i].GetValue().(type) {
> 		case *GetOperation:
> 			// TODO(arul): Looks like I forgot to add shared locks here.
> ```

Should we add this to the spreadsheet so we don't forget?
___
*[`pkg/kv/kvpb/api.go` line 1386 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net39BO9FvaX0ik-p5x:-NeybX2J3_jKBz_DKi84:bs5s5p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvpb/api.go#L1386)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Good point. Done.  

PS: (I think you inverted the condition above)
</blockquote></details>

Oh yeah, I meant to invert the condition, but then to have it be `if d == lock.Replicated { ... }`, so that this looks like `flagForLockStrength`. Also, I think we should change the name to `flagForLockDurability`.
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 150 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net4usc5hCjcHBn4HH6:-Neyc43-2jS4Q3uPovG8:brmbetp) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L150)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I was just clearing the entire thing because if there's a `LockConflictError` then none of the locks are acquired. Is that not so?
</blockquote></details>

I see, then I guess it would be `res.Local.AcquiredLocks = res.Local.AcquiredLocks[:0]`. I agree that if there's an error, no locks are acquired.
___
*[`pkg/kv/kvserver/client_replica_test.go` line 4992 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetdHqQAwtwfyz6yzsi:-NeydI5m4r7iivG-0a1f:be0yhai) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7664df138967f920b4401f64a180ac95ae742a4/pkg/kv/kvserver/client_replica_test.go#L4992)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I thought we'd need to wrap both the increments and `txn1.Commit()` and `txn2.Rollback()` inside of the mutex, because otherwise the `GetForUpdate` might get unblocked and read the wrong value for `numFinalized`. But that also makes this checking `numFinalized` bit kind of stupid. Should we just remove it?
</blockquote></details>

Yeah, I think we can remove it. Maybe we just double up the `expected GetForUpdate request to block` assertion and add it before and after `txn1.Commit`?
___
*[`pkg/kv/kvserver/client_replica_test.go` line 5018 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NeydTOd3CzBWZa4hhDL:-NeydTOd3CzBWZa4hhDM:b-zl8u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/671f4f4588568d4d46677be5ced1719e7324a22d/pkg/kv/kvserver/client_replica_test.go#L5018)):*
> ```Go
> 			require.NoError(t, err)
> 		case <-time.After(5 * time.Second):
> 			t.Fatal("timed out")
> ```

We can probably just rely on the test's timeout for this. So `require.NoError(t, <-ch)`. But also, I don't really get the purpose of this channel. Can we just assert `require.NoError(t, err)` in the goroutine above? We're already waiting for it to complete below with the wait group.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
RFAL! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-:-NeyrDHS-xsA-9XmNYTl:by76qkj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @miretskiy, and @nvanbenschoten)

___
*[`pkg/kv/kvnemesis/applier.go` line 289 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net6QyK3FvVSW6w-udQ:-Neymzw16Q82f6HxUcq-:b35j9n4) ([raw file](https://github.com/cockroachdb/cockroach/blob/639b8c8983627c5145c7ed9112310ccc56fe30c8/pkg/kv/kvnemesis/applier.go#L289)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we're assigning the method reference to a variable, then this avoids an allocation to close over `b`. But since we're now just calling the method directly, this is unnecessary.

We'll want to make this change down below as well.
</blockquote></details>

Thanks for the explanation. And done.
___
*[`pkg/kv/kvnemesis/applier.go` line 445 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Neycme66bbT4llmCaqK:-NeynOBX7QcQauAXPZ1s:bn7w3hm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cf34d617c8b3e63803bca897cb6da412ac613c0/pkg/kv/kvnemesis/applier.go#L445)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we add this to the spreadsheet so we don't forget?
</blockquote></details>

Added. I'll get to it as soon as this PR lands.
___
*[`pkg/kv/kvpb/api.go` line 1386 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net39BO9FvaX0ik-p5x:-Neyo2w34RnHp6P_7NvG:b-97593m) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvpb/api.go#L1386)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Oh yeah, I meant to invert the condition, but then to have it be `if d == lock.Replicated { ... }`, so that this looks like `flagForLockStrength`. Also, I think we should change the name to `flagForLockDurability`.
</blockquote></details>

Renamed, and done.
___
*[`pkg/kv/kvserver/batcheval/intent.go` line 150 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-Net4usc5hCjcHBn4HH6:-NeynaYV2o0jDVn0vKTj:br4cd4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/242dd69433773937452e4cadcfc92a28e4e0ed45/pkg/kv/kvserver/batcheval/intent.go#L150)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I see, then I guess it would be `res.Local.AcquiredLocks = res.Local.AcquiredLocks[:0]`. I agree that if there's an error, no locks are acquired.
</blockquote></details>

D'oh. 
___
*[`pkg/kv/kvserver/client_replica_test.go` line 4992 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NetdHqQAwtwfyz6yzsi:-NeypNC90guxXOhfxh9A:br2f38f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7664df138967f920b4401f64a180ac95ae742a4/pkg/kv/kvserver/client_replica_test.go#L4992)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yeah, I think we can remove it. Maybe we just double up the `expected GetForUpdate request to block` assertion and add it before and after `txn1.Commit`?
</blockquote></details>

Okay, done.
___
*[`pkg/kv/kvserver/client_replica_test.go` line 5018 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-NeydTOd3CzBWZa4hhDL:-NeypOX3EsK9W4OKOxOF:b28mmp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/671f4f4588568d4d46677be5ced1719e7324a22d/pkg/kv/kvserver/client_replica_test.go#L5018)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We can probably just rely on the test's timeout for this. So `require.NoError(t, <-ch)`. But also, I don't really get the purpose of this channel. Can we just assert `require.NoError(t, err)` in the goroutine above? We're already waiting for it to complete below with the wait group.
</blockquote></details>

Yeah, I was being fancy with this for no good reason. Simplified things.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 29 of 30 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110279#-:-Nf-1n4150hiacwKi_td:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @miretskiy)



<!-- Sent from Reviewable.io -->

arulajmani commented:
TFTR! 

bors r=nvanbenschoten

arulajmani commented:
bor r=nvanbenschoten

arulajmani commented:
bors r=nvanbenschoten

arulajmani commented:
`TestTransactionUnexpectedlyCommitted` seems to have flaked on both these runs. I've confirmed its a flake, and I'm gonna retry. 

bors r=nvanbenschoten

arulajmani commented:
I'm going to skip the test. https://github.com/cockroachdb/cockroach/pull/111205.

arulajmani commented:
Test is skipped now. Gonna merge. 

bors r=nvanbenschoten

arulajmani commented:
Things just timed out. I can't glean anything from the build logs that points to something with my patch, so I'll give it another go.

bors r=nvanbenschoten

arulajmani commented:
bors r-

arulajmani commented:
bors r+ p=99

arulajmani commented:
Day 4, we go again! 

bors r=nvanbenschoten
